### DISCUSSION TITLE
Efficiency of matrix assembly

### DISCUSSION THREAD

**Discussion by edljk** [Ferrite-FEM/Ferrite.jl#561] [General]
Is there a particular reason to motivate the quadratic cost in the assembling routines in several examples of the documentation? It seems to me significantly more efficient to replace (for instance)

```julia 
dΩ = getdetJdV(cellvalues, q_point)
for i in 1:n_basefuncs
    ∇v = shape_gradient(cellvalues, q_point, i)
    for j in 1:n_basefuncs
        ∇u = shape_gradient(cellvalues, q_point, j)
        Ke[i, j] += (∇v ⋅ ∇u) * dΩ
     end
 end
```
by something like

```julia 
dΩ = getdetJdV(cellvalues, q_point)
∇uv = [shape_gradient(cellvalues, q_point, k) for k in 1:n_basefuncs]
for i in 1:n_basefuncs
    ∇v = ∇uv[i]
    for j in 1:n_basefuncs
        ∇u = ∇uv[j]                        
        Ke[i, j] += (∇v ⋅ ∇u) * dΩ
    end
end
```
No? I am probably missing something.

**Comment by termi-official** [comment]
If you have the same space for your test and trial spaces, then you could do something along these lines. However, you will be likely slower, because the call to shape_gradient is already just a lookup.

**Comment by edljk** [comment]
In my very naive tests the gain is sometimes significant

**Comment by fredrikekre** [comment]
You can check https://ferrite-fem.github.io/Ferrite.jl/stable/examples/stokes-flow/ (https://github.com/Ferrite-FEM/Ferrite.jl/blob/570b3b5fa69d0ca56addf82dd9c8bbf790eec860/docs/src/literate/stokes-flow.jl#L386-L389 and https://github.com/Ferrite-FEM/Ferrite.jl/blob/570b3b5fa69d0ca56addf82dd9c8bbf790eec860/docs/src/literate/stokes-flow.jl#L397-L404) for an example where such caching is done. The important thing is that you allocate `∇uv` outside of the element routine, otherwise you pay the allocation cost for every element.

What did you measure? Just the assembly routine or the full global assembly?

**Comment by edljk** [comment]
Oups,  I missed that example. Yes, I know, my suggestion can be improved regarding allocation. This was just an illustration which has already an impact evaluating all the assembly:

* using the pre computing
```julia
assemb K           1    10.1s   86.7%   10.1s   3.94GiB   80.4%  3.94GiB
     ∇uv           402k    229ms    2.0%   570ns    178MiB    3.5%     464B
     comp dΩ       402k   60.7ms    0.5%   151ns   6.13MiB    0.1%    16.0B
     assemble!    5.50k   9.60ms    0.1%  1.74μs    344KiB    0.0%    64.0B
```

* initial implementation
```julia
 assemb K           1    21.4s   93.4%   21.4s   3.77GiB   79.6%  3.77GiB
     ∇u           40.2M    5.22s   22.8%   130ns   1.20GiB   25.3%    32.0B
     ∇v           4.02M    561ms    2.5%   140ns    123MiB    2.5%    32.0B
     comp dΩ       402k   51.4ms    0.2%   128ns   6.13MiB    0.1%    16.0B
     assemble!    5.50k   9.44ms    0.0%  1.71μs    344KiB    0.0%    64.0B
```

**Comment by fredrikekre** [comment]
That looks very suspicious. What are you using to measure that? In particular, `∇u` should be completely allocation free.

**Reply by koehlerson** [reply]
looks like TimerOutputs.jl

**Reply by edljk** [reply]
Yes it is

**Reply by fredrikekre** [reply]
Yea. Then I think if you put `@timeit`s in such a hot loop that will cause huge slowdowns.

**Reply by fredrikekre** [reply]
See e.g. https://github.com/KristofferC/TimerOutputs.jl#overhead

**Comment by edljk** [comment]
Great thanks, sorry for the noise..
I do not know if it could also contribute to explain the observations: I am using `P3` elements with second order boundary

**Reply by koehlerson** [reply]
could you `@btime` it maybe with `BenchmarkTools` I'm curious to see the comparison, but if its too much to change, then don't mind :)

**Comment by fredrikekre** [comment]
As a point of reference, for the heat equation with 200x200 elements, third order interpolation, assembly is slightly longer when caching the values (140ms without caching, 155ms with caching). Likely because you only shuffle data around with the caching, but then have to do almost the equivalent lookup later anyway (i.e. accessing `∇v = ∇uv[i]`).

As pointed out above, `shape_value` and `shape_gradient` is just a lookup (See https://github.com/Ferrite-FEM/Ferrite.jl/blob/570b3b5fa69d0ca56addf82dd9c8bbf790eec860/src/FEValues/common_values.jl#L84 and https://github.com/Ferrite-FEM/Ferrite.jl/blob/570b3b5fa69d0ca56addf82dd9c8bbf790eec860/src/FEValues/common_values.jl#L96). This might change if you use e.g. `shape_divergence` or `shape_symmetric_gradient` which does some very trivial computation too. (See https://github.com/Ferrite-FEM/Ferrite.jl/blob/570b3b5fa69d0ca56addf82dd9c8bbf790eec860/src/FEValues/common_values.jl#L114 and https://github.com/Ferrite-FEM/Ferrite.jl/blob/570b3b5fa69d0ca56addf82dd9c8bbf790eec860/src/FEValues/common_values.jl#L105). However, compared to assembling into the stiffness matrix, or even solving the linear system, I don't expect the caching to be very beneficial in many cases. But as always, benchmarking is your friend :D

**Reply by termi-official** [reply]
I want to leave a reference from Prof. Wells here, whose group has done some benchmarking on this stuff over the recent decade. https://youtu.be/D-YcVd4-_2E?t=2550 . In their analyses they concluded that for higher order elements we start running into trouble with too  much caching, because of earlier cache spilling.

**Comment by edljk** [comment]
Thank you so much for the reactivity.
Yes, it seems using `@timeit` in an irrelevant way is the explanation. 
Using `btime` with the following peaces of codes (not completely sure it is the right way), I can not observe a significant difference anymore

* using the pre computing
```julia 
@btime begin
for k in 1:$n_basefuncs
     $∇uv[k] = shape_gradient($cellvalues, $q_point, k) 
end           
for i in 1:$n_basefuncs
    ∇v = $∇uv[i]
    for j in 1:$n_basefuncs
         u = $∇uv[j]                        
         $Ke[i, j] += (∇v ⋅ ∇u) * $dΩ
     end
end
end

  139.512 ns (0 allocations: 0 bytes)
  111.476 ns (0 allocations: 0 bytes)
  142.031 ns (0 allocations: 0 bytes)
  111.493 ns (0 allocations: 0 bytes)
  139.515 ns (0 allocations: 0 bytes)
  111.469 ns (0 allocations: 0 bytes)
  142.025 ns (0 allocations: 0 bytes)
  111.829 ns (0 allocations: 0 bytes)
...
```

* initial implementation
```julia
@btime begin         
for i in 1:$n_basefuncs
     ∇v = shape_gradient($cellvalues, $q_point, i)
      for j in 1:$n_basefuncs
           ∇u = shape_gradient($cellvalues, $q_point, j)
           $Ke[i, j] += (∇v ⋅ ∇u) * $dΩ
      end
end
end

  116.681 ns (0 allocations: 0 bytes)
  150.695 ns (0 allocations: 0 bytes)
  118.124 ns (0 allocations: 0 bytes)
  149.268 ns (0 allocations: 0 bytes)
  117.144 ns (0 allocations: 0 bytes)
  150.820 ns (0 allocations: 0 bytes)
  115.594 ns (0 allocations: 0 bytes)
  149.268 ns (0 allocations: 0 bytes)
...

```
